File: hw/zehuaZhang/hw1bullsEye/techbowA1objC/techbowA1objC/ViewController.m
//
//  ViewController.m
//  techbowA1objC
//
//  Created by ZhangZehua on 11/12/15.
//  Copyright © 2015 ZhangZehua. All rights reserved.
//

#import "ViewController.h"

@interface ViewController ()
@property (nonatomic) NSInteger currentValue;
@property (nonatomic) NSInteger targetValue;
@property (nonatomic) NSInteger score;
@property (nonatomic) NSInteger round;


@end

@implementation ViewController

- (instancetype) initWithCoder:(NSCoder *)aDecoder {
    if (self = [super initWithCoder:aDecoder]) {
//        _currentValue = 50;
//        _targetValue = 0;
//        _score = 0;
//        _round = 0;
    }
    return self;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view, typically from a nib.
    
    [self startNewGame];
    [self updateLabels];
    
    UIImage *thumbImageNormal = [UIImage imageNamed:@"SliderThumb-Normal"];
    [_slider setThumbImage:thumbImageNormal forState:UIControlStateNormal];
    
    UIImage *thumbImageHighlighted = [UIImage imageNamed:@"SliderThumb-Highlighted"];
    [_slider setThumbImage:thumbImageHighlighted forState:UIControlStateHighlighted];
    
    UIEdgeInsets insets = UIEdgeInsetsMake(0, 14, 0, 14);
    
    UIImage *trackLeftImage = [UIImage imageNamed:@"SliderTrackLeft"];
    UIImage *trackRightImage = [UIImage imageNamed:@"SliderTrackRight"];
    if (trackLeftImage != nil) {
        UIImage *trackLeftResizable = [trackLeftImage resizableImageWithCapInsets:insets];
        [_slider setMinimumTrackImage:trackLeftResizable forState:UIControlStateNormal];
    }
    if (trackRightImage != nil) {
        UIImage *trackRightResizable = [trackRightImage resizableImageWithCapInsets:insets];
        [_slider setMaximumTrackImage:trackRightResizable forState:UIControlStateNormal];
    }
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (IBAction) startOver:(UIButton *)sender {
    [self startNewGame];
    [self updateLabels];
    
    CATransition *transition = [CATransition animation];
    transition.type = kCATransitionFade;
    transition.duration = 0.5;
    transition.timingFunction = [CAMediaTimingFunction functionWithName:kCAMediaTimingFunctionEaseOut];
    [self.view.layer addAnimation:transition forKey: nil];
    
}

- (IBAction) showAlert {
    NSInteger difference = (NSInteger) labs(_targetValue - _currentValue);
    NSInteger points = 100 - difference;
    NSString *title;
    
    if (difference == 0) {
        title = @"Perfect!";
        points += 100;
    } else if (difference < 5) {
        title = @"You almost had it!";
        if (difference == 1) {
            points += 50;
        }
    } else if (difference < 10) {
        title = @"Pretty good!";
    } else {
        title = @"Not even close...";
    }
    
    _score += points;
    
    NSString *message = [NSString stringWithFormat:@"You scored %ld points", (long)points];
    UIAlertController *alert = [UIAlertController alertControllerWithTitle: title message: message preferredStyle: UIAlertControllerStyleAlert];
    UIAlertAction *action = [UIAlertAction actionWithTitle:title style:UIAlertActionStyleDefault
                                            handler:^(UIAlertAction * _Nonnull action) {
        [self startNewRound];
        [self updateLabels];
    }];
    [alert addAction:action];
    [self presentViewController:alert animated:true completion:nil];
}

- (IBAction)sliderMoved:(UISlider *)sender {
    _currentValue = lroundf(sender.value);
}

- (void) startNewRound {
    _round++;
    _targetValue = 1 + (NSInteger) arc4random_uniform(100);
    _currentValue = 50;
    _slider.value = (float) _currentValue;
}

- (void) startNewGame {
    _score = 0;
    _round = 0;
    [self startNewRound];
}

- (void) updateLabels {
    _targetLabel.text = [@(_targetValue) stringValue];
    _scoreLabel.text = [@(_score) stringValue];
    _roundLabel.text = [@(_round) stringValue];
}

@end

File: hw/zehuaZhang/hw1bullsEye/techbowA1/techbowA1/ViewController.swift
//
//  ViewController.swift
//  techbowA1
//
//  Created by ZhangZehua on 11/10/15.
//  Copyright © 2015 ZhangZehua. All rights reserved.
//

import UIKit
import QuartzCore

class ViewController: UIViewController {

    var currentValue = 50
    var targetValue = 0
    var score = 0
    var round = 0
    
    @IBOutlet weak var targetLabel: UILabel!
    @IBOutlet weak var slider: UISlider!
    @IBOutlet weak var scoreLabel: UILabel!
    @IBOutlet weak var roundLabel: UILabel!
    
    @IBAction func startOver(sender: UIButton) {
        startNewGame()
        updateLabels()
        
        let transition = CATransition()
        transition.type = kCATransitionFade
        transition.duration = 0.5
        transition.timingFunction = CAMediaTimingFunction(name: kCAMediaTimingFunctionEaseOut)
        view.layer.addAnimation(transition, forKey: nil)
    }

    @IBAction func showAlert() {
        let difference = abs(targetValue - currentValue)
        var points = 100 - difference

        var title: String
        if difference == 0 {
            title = "Perfect!"
            points += 100
        } else if difference < 5{
            title = "You almost had it!"
            if difference == 1 {
                points += 50
            }
        } else if difference < 10 {
            title = "Pretty good!"
        } else {
            title = "Not even close..."
        }
        
        score += points
        
        let message = "You scored \(points) points"
        let alert = UIAlertController(title: title, message: message, preferredStyle: .Alert)
        let action = UIAlertAction(title: "OK", style: .Default, handler: {
            act in
            self.startNewRound()
            self.updateLabels()

        })
        alert.addAction(action)
        presentViewController(alert, animated: true, completion: nil)
    }
    
    @IBAction func sliderMoved(sender: UISlider) {
        currentValue = lroundf(sender.value)
        //print("The value of the slider is now: \(sender.value)")
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
        startNewGame()
        updateLabels()
        
        let thumbImageNormal = UIImage(named: "SliderThumb-Normal")
        slider.setThumbImage(thumbImageNormal, forState: .Normal)
        
        let thumbImageHighlighted = UIImage(named: "SliderThumb-Highlighted")
        slider.setThumbImage(thumbImageHighlighted, forState: .Highlighted)
        
        let insets = UIEdgeInsets(top: 0, left: 14, bottom: 0, right: 14)
        
        if let trackLeftImage = UIImage(named: "SliderTrackLeft") {
            let trackLeftResizable = trackLeftImage.resizableImageWithCapInsets(insets)
            slider.setMinimumTrackImage(trackLeftResizable, forState: .Normal)
        }
        
        if let trackRightImage = UIImage(named: "SliderTrackRight") {
            let trackRightResizable = trackRightImage.resizableImageWithCapInsets(insets)
            slider.setMaximumTrackImage(trackRightResizable, forState: .Normal)
        }
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    func startNewRound() {
        round++
        targetValue = 1 + Int(arc4random_uniform(100))
        currentValue = 50
        slider.value = Float(currentValue)
    }
    func updateLabels() {
        targetLabel.text = String(targetValue)
        scoreLabel.text = "\(score)"
        roundLabel.text = "\(round)"
        //targetLabel.text = "\(targetValue)"
    }
    func startNewGame() {
        score = 0
        round = 0
        startNewRound()
    }
}

